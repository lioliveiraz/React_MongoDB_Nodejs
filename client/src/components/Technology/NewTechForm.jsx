import React, { useState, useEffect } from "react";
import Input from "./../Base/Input";
import Button from "@material-ui/core/Button";
import { addNewTech } from "../../store/actions/wall";
import { connect } from "react-redux";
import { getCategories } from "./../../store/actions/categories";
import { useHistory } from "react-router-dom";
import { toast } from "react-toastify";
import techValidator from "../../helpers/validations/techValidator";
import { InputLabel, Paper, Select, TextareaAutosize } from "@material-ui/core";
import MenuItem from "@material-ui/core/MenuItem";
import { Grid } from "@material-ui/core";
import { useStyles } from "./../../assets/css/Technologies/newTechForm";
import Typography from "@material-ui/core/Typography";

function NewTechForm({ addNewTech, token, getCategories, categories, wall }) {
  const [techObject, setTechObject] = useState({});
  const [category, setCategory] = useState("");
  const [errors, setErrors] = useState({});
  const history = useHistory();
  const classes = useStyles();

  useEffect(() => {
    getCategories();
  }, [getCategories]);

  const getUserInput = (value, name) => {
    setTechObject({
      ...techObject,
      [name]: value,
    });
  };

  const submit = async (e) => {
    e.preventDefault();
    const inputErrors = techValidator(techObject);
    if (inputErrors) {
      setErrors(inputErrors);
      toast.error("Something is not right!");
    } else {
      addNewTech(history, techObject, token);
    }
  };

  useEffect(() => {
    if (wall.errors.length > 0) {
      return wall.errors.forEach(({ msg }) => toast.error(msg));
    }
    if (wall.status) {
      return toast.success(wall.status);
    }
  }, [wall.errors, wall.status]);

  return (
    <Grid item xs={10} className={classes.form}>
      <Paper elevation={2} className={classes.paper}>
        <Typography variant="h3" color="primary" className={classes.title}>
          {" "}
          New Tech
        </Typography>

        <Grid item>
          <form>
            <Grid item className={classes.input}>
              <Input
                type="text"
                name="name"
                required
                placeholder="Tech name"
                getUserInput={getUserInput}
                error={errors.name && errors.name}
              />
            </Grid>
            <Grid item className={classes.input}>
              <TextareaAutosize
                name="description"
                cols="30"
                rows="10"
                placeholder="Description"
                onChange={(e) => getUserInput(e.target.value, e.target.name)}
              />
              {errors.description && errors.description}
              <InputLabel id="demo-simple-select-label">{category}</InputLabel>
            </Grid>
            <Grid item className={classes.input}>
              <InputLabel id="demo-simple-select-label">
                Select Category
              </InputLabel>

              <Select
                labelId="demo-simple-select-label"
                name="category"
                className={classes.select}
                value={category}
                onChange={(e) => {
                  getUserInput(e.target.value, e.target.name);
                  setCategory(e.target.name);
                }}
              >
                {categories &&
                  categories.map(({ _id, name, color }) => (
                    <MenuItem
                      key={name}
                      style={{ background: color, color: "white" }}
                      value={_id}
                    >
                      {name}
                    </MenuItem>
                  ))}
              </Select>
            </Grid>
            <Grid item className={classes.input}>
              <Input
                type="text"
                name="creator"
                required
                placeholder="Creator"
                getUserInput={getUserInput}
                error={errors.creator && errors.creator}
              />
            </Grid>

            <Button
              onClick={submit}
              variant="contained"
              color="secondary"
              disabled={techObject.name && techObject.category ? false : true}
            >
              {" "}
              Submit
            </Button>
          </form>
        </Grid>
      </Paper>
    </Grid>
  );
}

const mapStateToProps = (state) => {
  return {
    categories: state.category.categories,
    token: state.auth.token,
    wall: state.wall,
  };
};

export default connect(mapStateToProps, { addNewTech, getCategories })(
  NewTechForm
);
