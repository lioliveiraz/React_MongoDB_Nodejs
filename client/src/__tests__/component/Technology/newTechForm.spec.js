import React from "react";
import { mount, shallow, spyOn } from "enzyme";
import configureStore from "redux-mock-store";
import { BrowserRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";
import NewTechForm from "../../../components/Technology/NewTechForm";
import thunk from "redux-thunk";

const middleware = [thunk];

let initialState = {
  category: {
    categories: [
      { id: 1, name: "framework", color: "yellow" },
      { id: 2, name: "framework", color: "yellow" },
      { id: 3, name: "framework", color: "yellow" },
    ],
  },
  auth: {
    token: "token",
  },
  wall: {
    pool: [],
    warm: [],
    hot: [],
    cold: [],
    errors: [],
  },
};

describe("SignUp", () => {
  let wrapper;
  const mockStore = configureStore(middleware);
  const store = mockStore(initialState);
  const historyMock = { push: jest.fn() };

  beforeEach(() => {
    wrapper = mount(
      <Router>
        <Provider store={store}>
          <NewTechForm history={historyMock} />
        </Provider>
      </Router>
    );
  });
  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
