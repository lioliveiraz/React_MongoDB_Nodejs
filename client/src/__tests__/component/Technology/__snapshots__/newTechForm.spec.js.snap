// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should match snapshot 1`] = `ReactWrapper {}`;

exports[`SignUp should match snapshot 2`] = `"<form><div class=\\"MuiFormControl-root MuiTextField-root makeStyles-root-1\\" data-cy=\\"input-name-wrapper\\"><label class=\\"MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"standard-secondary\\" id=\\"standard-secondary-label\\">Name<span aria-hidden=\\"true\\" class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiOutlinedInput-colorSecondary MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"standard-secondary\\" name=\\"name\\" placeholder=\\"Tech name\\" required=\\"\\" type=\\"text\\" data-cy=\\"input-name\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-4\\"><span>Name&nbsp;*</span></legend></fieldset></div></div><textarea rows=\\"10\\" name=\\"description\\" cols=\\"30\\" placeholder=\\"Description\\" style=\\"height: -4px; overflow: hidden;\\"></textarea><textarea aria-hidden=\\"true\\" readonly=\\"\\" tabindex=\\"-1\\" style=\\"visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0);\\"></textarea><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated\\" id=\\"demo-simple-select-label\\">Select Category</label><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline\\"><div class=\\"MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input\\" tabindex=\\"0\\" role=\\"button\\" aria-haspopup=\\"listbox\\" aria-labelledby=\\"demo-simple-select-label mui-component-select-category\\" id=\\"mui-component-select-category\\"><span>​</span></div><input name=\\"category\\" aria-hidden=\\"true\\" tabindex=\\"-1\\" class=\\"MuiSelect-nativeInput\\" value=\\"Select Category\\"><svg class=\\"MuiSvgIcon-root MuiSelect-icon\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M7 10l5 5 5-5z\\"></path></svg></div><div class=\\"MuiFormControl-root MuiTextField-root makeStyles-root-1\\" data-cy=\\"input-creator-wrapper\\"><label class=\\"MuiFormLabel-root MuiFormLabel-colorSecondary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required\\" data-shrink=\\"false\\" for=\\"standard-secondary\\" id=\\"standard-secondary-label\\">Creator<span aria-hidden=\\"true\\" class=\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"> *</span></label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorSecondary MuiOutlinedInput-colorSecondary MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"standard-secondary\\" name=\\"creator\\" placeholder=\\"Creator\\" required=\\"\\" type=\\"text\\" data-cy=\\"input-creator\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-4\\"><span>Creator&nbsp;*</span></legend></fieldset></div></div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-text\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiButton-label\\"> Submit</span><span class=\\"MuiTouchRipple-root\\"></span></button></form>"`;
